interactions/api/gateway/client.py

```diff
@@ -646,32 +688,109 @@ def __option_type_context(self, context: "_Context", type: int) -> dict:
+
+            self._client = await self._http._req._session.ws_connect(url, **self._options)
+
+            data = await self.__receive_packet(True)  # First data is the hello packet.
+
+            self.__heartbeater.delay = data["d"]["heartbeat_interval"]
+
+            if self.__task:
+                self.__task.cancel()
+                if self.__heartbeat_event.is_set():
+                    self.__heartbeat_event.clear()  # Because we're hardresetting the process
+
+                self.__task = create_task(self.run_heartbeat())
+
+            if not to_resume:
+                await self.__identify(self.__shard, self.__presence)
+            else:
+                await self.__resume()
+
+            self.__closed.set()
+            self.__heartbeat_event.set()
+
+    async def __receive_packet(self, ignore_lock: bool = False) -> Optional[Dict[str, Any]]:
         """
-        Receives a stream of packets sent from the Gateway.
+        Receives a stream of packets sent from the Gateway in an async process.
 
         :return: The packet stream.
         :rtype: Optional[Dict[str, Any]]
         """
 
-        packet: WSMessage = await self._client.receive()
+        while True:
+
+            if not ignore_lock:
+                # meaning if we're reconnecting or something because of tasks
+                await self.__closed.wait()
+
+            packet: WSMessage = await self._client.receive()
+
+            if packet.type == WSMsgType.CLOSE:
+                log.debug(f"Disconnecting from gateway = {packet.data}::{packet.extra}")
+
+                if packet.data >= 4000:
+                    # This means that the error code is 4000+, which may signify Discord-provided error codes.
+                    raise LibraryException(packet.data)  # Works because gateway!!
+                    # We need to work on __aexit__ / __aentry__ though, since we're not using the ws request manager.
+
+                if ignore_lock:
+                    raise LibraryException(
+                        message="Discord unexpectedly wants to close the WS on receiving by force.",
+                        severity=50,
+                    )
+
+                await self._reconnect(packet.data != 1000, packet.data)
+                continue
 
-        if packet == WSMsgType.CLOSE:
-            await self._client.close()
-            return packet
+            elif packet.type == WSMsgType.CLOSED:
+                # We need to wait/reconnect depending about other event holders.
 
-        elif packet == WS_CLOSED_MESSAGE:
-            return packet
+                if ignore_lock:
+                    raise LibraryException(
+                        message="Discord unexpectedly closed on receiving by force.", severity=50
+                    )
+
+                if not self.__closed.is_set():
+                    await self.__closed.wait()
+
+                    # Edge case on force reconnecting if we dont
+                else:
```

Comment:
So you don't want to reconnect if closed is not set?

Reply:
Race conditions/edge cases tbh, as closed is set on connect/reconnect